### PR TITLE
Pass netrc data to external config service request

### PR DESCRIPTION
**Pull request by pinpox**
As discussed in https://github.com/woodpecker-ci/woodpecker/discussions/2268#discussioncomment-6777977

closes #2268

### CODE REVIEW

File: server/forge/configFetcher.go [hunk 1]
```diff
@@ -74,7 +74,14 @@ func (cf *configFetcher) Fetch(ctx context.Context) (files []*types.FileMeta, er
 			defer cancel() // ok here as we only try http fetching once, returning on fail and success
 
 			log.Trace().Msgf("ConfigFetch[%s]: getting config from external http service", cf.repo.FullName)
-			newConfigs, useOld, err := cf.configExtension.FetchConfig(fetchCtx, cf.repo, cf.pipeline, files)
+			netrc, err := cf.forge.Netrc(cf.user, cf.repo)
+
+			if err != nil {
+				log.Error().Msg("Failed to get netrc " + err.Error())
+				return nil, fmt.Errorf("On Fetching config via http : %w", err)
+			}
```
Comment (anbraten) [inline server/forge/configFetcher.go:82]:
```suggestion
			if err != nil {
				log.Error().Err(err).Msg("could not get Netrc data from forge")
				return nil, fmt.Errorf("could not get Netrc data from forge: %w", err)
			}
```

File: server/forge/configFetcher.go [hunk 2]
```diff
@@ -74,7 +74,14 @@ func (cf *configFetcher) Fetch(ctx context.Context) (files []*types.FileMeta, er
...
+
+			newConfigs, useOld, err := cf.configExtension.FetchConfig(fetchCtx, cf.repo, cf.pipeline, files, netrc)
 			if err != nil {
 				log.Error().Msg("Got error " + err.Error())
 				return nil, fmt.Errorf("On Fetching config via http : %w", err)
```
Comment (anbraten) [inline server/forge/configFetcher.go:87]:
```suggestion
				log.Error().Err(err).Msg("could not fetch config via http")
				return nil, fmt.Errorf("could not fetch config via http: %w", err)
```

**Review by anbraten — APPROVED** [review]

File: server/forge/configFetcher.go [hunk 3]
```diff
@@ -74,12 +74,19 @@ func (cf *configFetcher) Fetch(ctx context.Context) (files []*types.FileMeta, er
 			defer cancel() // ok here as we only try http fetching once, returning on fail and success
 
 			log.Trace().Msgf("ConfigFetch[%s]: getting config from external http service", cf.repo.FullName)
-			newConfigs, useOld, err := cf.configExtension.FetchConfig(fetchCtx, cf.repo, cf.pipeline, files)
+			netrc, err := cf.forge.Netrc(cf.user, cf.repo)
+
 			if err != nil {
-				log.Error().Msg("Got error " + err.Error())
+				log.Error().Msg("Failed to get netrc " + err.Error())
 				return nil, fmt.Errorf("On Fetching config via http : %w", err)
```
Comment (anbraten) [inline server/forge/configFetcher.go:81]:
```suggestion
				log.Error().Err(err).Msg("could not get Netrc data from forge")
				return nil, fmt.Errorf("could not get Netrc data from forge: %w", err)
```

File: server/forge/configFetcher.go [hunk 4]
```diff
@@ -74,10 +74,17 @@ func (cf *configFetcher) Fetch(ctx context.Context) (files []*types.FileMeta, er
 			defer cancel() // ok here as we only try http fetching once, returning on fail and success
 
 			log.Trace().Msgf("ConfigFetch[%s]: getting config from external http service", cf.repo.FullName)
-			newConfigs, useOld, err := cf.configExtension.FetchConfig(fetchCtx, cf.repo, cf.pipeline, files)
+			netrc, err := cf.forge.Netrc(cf.user, cf.repo)
+
```
Comment (anbraten) [inline server/forge/configFetcher.go:78]:
```suggestion
```

**Comment by pinpox** [comment]
A mock for the tests was missing. I havn't worked with this mocking library before, but would assume form the other examples that this is how it is supposed to be added.

**Comment by 6543** [comment]
can we add a note to the docs that the external service must be trustworthy as it gets the netrc ?

**Comment by pinpox** [comment]
> can we add a note to the docs that the external service must be trustworthy as it gets the netrc ?

Added to the external configuration servcie docs

**Comment by anbraten** [comment]
> can we add a note to the docs that the external service must be trustworthy as it gets the netrc ?

Not only netrc, but also the config and the power to change the config to anything 😅 

A red box on this page would be nice:

https://woodpecker-ci.org/docs/administration/external-configuration-api

:::warning
You need to trust the external config service as it is getting secret information about the repository and pipeline and has the ability to change pipeline configs that could run malicious tasks.
:::

**Comment by pinpox** [comment]
Updated to the text above with box.

**Review by anbraten — APPROVED** [review]

File: server/api/pipeline.go [hunk 5]
```diff
@@ -453,7 +453,12 @@ func PostPipeline(c *gin.Context) {
 		}
 	}
 
-	newpipeline, err := pipeline.Restart(c, _store, pl, user, repo, envs)
+	netrc, err := server.Config.Services.Forge.Netrc(user, repo)
+	if err != nil {
+		handlePipelineErr(c, err)
+	}
```
[the remark this answers is not in the document]
Comment (pinpox) [reply server/api/pipeline.go:459]: Looking at the code above again, should I maybe use 
```
_ = c.AbortWithError(http.StatusBadRequest, err)
return
```

instaed of `handlePipelineErr(c, err)` ? Or even use `http.StatusUnauthorized`?
If you let me know what error handling you prefer, I can make a PR to fix it

[the remark this answers is not in the document]
Comment (6543) [reply server/api/pipeline.go:459]: thanks :)

**Pull request merged**